src/module.c

```diff
@@ -3048,6 +3087,227 @@ int RM_HashGet(RedisModuleKey *key, int flags, ...) {
     return REDISMODULE_OK;
 }
 
+/* --------------------------------------------------------------------------
+ * Key API for the stream type.
+ * -------------------------------------------------------------------------- */
+
+/* Adds an entry to a stream. Like XADD without trimming.
+ *
+ * - `key`: The key where the stream is (or will be) stored
+ * - `flags`: A bit field of
+ *   - `REDISMODULE_STREAM_AUTOID`: Assign a stream ID automatically, like `*`
+ *     in the XADD command.
+ *   - `REDISMODULE_STREAM_NOMKSTREAM`: Don't create the stream if it doesn't
+ *     exist.
+ * - `id`: If the `AUTOID` flag is set, this is where the assigned ID is
+ *   returned. Can be NULL if `AUTOID` is set, if you don't care to receive the
+ *   ID. If `AUTOID` is not set, this is the requested ID.
+ * - `argv`: A pointer to an array of size `numfields * 2` containing the
+ *   fields and values.
+ * - `numfields`: The number of field-value pairs in `argv`.
+ *
+ * Returns REDISMODULE_OK if an entry has been added. Returns REDISMODULE_ERR if
+ * the key is not opened for writing, if the key has a type other than stream,
+ * if the key doesn't exist in case the flag `NOMKSTREAM` is set, if unknown
+ * flags are given or when attempting to create an entry with an id which is not
+ * allowed.
+ */
+int RM_StreamAdd(RedisModuleKey *key, int flags, RedisModuleStreamID *id, RedisModuleString **argv, long numfields) {
+    /* Validate args */
+    if (flags & ~(REDISMODULE_STREAM_AUTOID |
+                  REDISMODULE_STREAM_NOMKSTREAM))
+        return REDISMODULE_ERR; /* unknown flags */
+    if (!(key->mode & REDISMODULE_WRITE))
+        return REDISMODULE_ERR;
+    if (key->value && key->value->type != OBJ_STREAM)
+        return REDISMODULE_ERR;
+    if (key->value == NULL && (flags & REDISMODULE_STREAM_NOMKSTREAM))
+        return REDISMODULE_ERR;
+    if (!(flags & REDISMODULE_STREAM_AUTOID) &&
+        (id == NULL || (id->ms == 0 && id->seq == 0)))
+        return REDISMODULE_ERR;
+
+    /* Create key if necessery */
+    if (key->value == NULL) {
+        moduleCreateEmptyKey(key, REDISMODULE_KEYTYPE_STREAM);
+    }
+
+    stream *s = key->value->ptr;
+    streamID added_id;
+    streamID use_id;
+    streamID *use_id_ptr = NULL;
+    if (!(flags & REDISMODULE_STREAM_AUTOID)) {
+        use_id.ms = id->ms;
+        use_id.seq = id->seq;
+        use_id_ptr = &use_id;
+    }
+    if (streamAppendItem(s, argv, numfields, &added_id, use_id_ptr) == C_ERR) {
+        /* Current last ID is greater than or equal to 'use_id' */
+        return REDISMODULE_ERR;
+    }
+
+    if (id != NULL) {
+        id->ms = added_id.ms;
+        id->seq = added_id.seq;
+    }
+
+    return REDISMODULE_OK;
+}
+
+/* Sets up a stream iterator.
+ *
+ * - `key`: The stream key opened for reading using RedisModule_OpenKey().
+ * - `flags`:
+ *   - `REDISMODULE_STREAM_EXCLUSIVE`: Don't include `start` and `end` in the
```

Comment:
maybe instead of the `EXCLUSIVE` flag we better just expose streamIncrID and streamDecrID as APIs?
it would be a bit harder to use, but can be more powerful.
i'm not sure.
@guybe7 @itamarhaber ?

Reply:
Give me a use case for incr and decr and I will be convinced. :-)